Microsoft.Toolkit/Microsoft.Toolkit.csproj

```diff
@@ -14,4 +16,54 @@
     <DebugType>Full</DebugType>
   </PropertyGroup>
 
+  <ItemGroup>
+    <PackageReference Include="System.Memory" Version="4.5.3" />
```

Comment:
It's for a couple reasons:

- The `Guard` APIs has a number of overloads that support `Span<T>`, `ReadOnlySpan<T>`, `Memory<T>` and `ReadOnlyMemory<T>`, yeah.
- The package also includes the `System.Runtime.CompilerServices.Unsafe` class, which is used by a number of APIs I've added (in particular, `Guard.IsBitwiseEqualTo<T>` and `ToHexString<T>`).

All the new `Span<T>`-related APIs are also available by default both in .NET Standard 2.1 and .NET Core 3.0, so eg. if we were to eventually multitarget the `Microsoft.Toolkit` package, we wouldn't even need to reference `System.Memory` directly (just `System.Runtime.CompilerServices.Unsafe` at that point).

In general, I'd say `System.Memory` is absolutely a must have package for everyone: using `Span<T>`-related APIs should **always** be preferred over creating/copying new arrays, as the speed and memory improvements are just incredible in most cases.

In fact, after this PR is merged I'd like to change a couple of extensions from the `ArrayExtensions` class to return `Memory<T>` instead of `T[]`, as it'd make them allocation-free and much more efficient. And we should definitely also consider adding overloads to those APIs (eg. `ToArrayString`) to support `Span<T>` as well, it'd make them much more flexible.

Of course I'd be happy to help with that as well 😊

Reply:
Cool, FYI if we change the signatures to the ArrayExtensions, we'll need to do that work for the 7.0 release as it'd be a breaking change, eh?

Reply:
Yeah absolutely, that's one of the reasons why I didn't touch those APIs in this PR 😊

Reply:
@michael-hawker, I don't know if there is a tool we can use to validate the impact. 
We may be able to use the size of the sample app as a reference value (even if not 100% accurate) but I have no idea on how to do that 🙁.

Reply:
Thanks @vgromfeld, no worries, never hurts to ask. :)

Reply:
I realize it's probably the best comparison, but just as a test I tried to build a blank UWP app in 3 configurations:

1. Just the app with no references. I just added a test method that simply creates a random array, loops over it and prints the values, just to have some conditions. I used explicit `throw` statements.
2. I referenced `Microsoft.Toolkit` **without** the `System.Memory` package, and replaced those `throw` expressions with `Guard` calls.
3. I referenced the current `Microsoft.Toolkit` package as in this PR.

![image](https://user-images.githubusercontent.com/10199417/76576502-8a99f500-64c2-11ea-82cf-b4b79aab2267.png)

Then I created a package for each version in Release x64 (so with .NET Native).
You can see that the difference in the final package is just about **20KB** with/without the `System.Memory` package. Which makes sense, the package itself is fairly small and only contains a few types and classes, not much else.

Hope this helps! 😊

Reply:
Thanks @Sergio0694 and to clarify was #2 with the Microsoft.Toolkit with this PR or without the PR? (Curious about the PRs impact on the bottom-line anyway as we're having this discussion.)

Reply:
\#2 was with the same exact `Microsoft.Toolkit` package as this PR, just after removing the `System.Memory` reference and all the APIs that were using types from that package (eg. all the `Guard` APIs working on `Span<T>`). I did this to highlight the difference of just the `System.Memory` package, as opposed to also including the binary size increase just caused by the new `Guard` APIs and the other extensions.

Reply:
As per @michael-hawker's request, here's a text breakdown of the binary size in my test application (UWP, Release x64, SDK 18363):

1) **No packages:**  `24.604.285` byte
2) **`Guard` API, direct reference:** `24.710.254` byte
3) **`Guard` API, direct, no `System.Memory`:** `24.690.486` byte
4) **`Microsoft.Toolkit` from NuGet:** `24.620.377` byte

Looks like the final binary size in a UWP application is actually less when the package is referenced from NuGet as opposed to when the project is referenced directly from the Visual Studio solution. And in 4) the final size was actually smaller than both the versions of this PR, with and without the `System.Memory` package being referenced.